opencog/rule-engine/Rule.cc

```diff
@@ -484,13 +481,23 @@ std::string Rule::to_string(const std::string& indent) const
 	return ss.str();
 }
 
-Rule Rule::rand_alpha_converted() const
+Rule Rule::rand_alpha_converted(const Handle& term, const Handle& vardecl)
+const
 {
+	const Handle& m_vardecl = gen_vardecl(term, vardecl);
 	// Clone the rule
 	Rule result = *this;
 
-	// Alpha convert the rule
-	result.set_rule(_rule->alpha_convert());
+	HandleSet target_vars = VariableListCast(m_vardecl)->get_variables().varset;
+	HandleSet result_vars = BindLinkCast(result.get_rule())->get_variables()
+			.varset;
```

Comment:
You may directly use `result._rule` instead of casting `result.get_rule()`, that is because a class is automatically a friend of itself, see https://stackoverflow.com/questions/39404109/is-a-class-automatically-a-friend-of-itself

Also, better have a const reference of HandleSet to not waste CPU copying varset.